### PR TITLE
Upload Service refactoring

### PR DESCRIPTION
**Pull request by Abhinav-grd**

### CODE REVIEW

File: src/utils/upload/index.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import { FileWithCollection } from 'services/upload/uploadManager';
+import { FileInMemory, MetadataObject } from 'services/upload/uploadService';
+
+const TYPE_JSON = 'json';
+
+export function fileAlreadyInCollection(
+    existingFilesInCollection,
```
Comment (ua741) [inline src/utils/upload/index.ts:7]: Is it possible to specify type of `existingFilesInCollection` similar to `newFile:` **FileInMemory**?

File: src/utils/network/index.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { sleep } from 'utils/common';
+
+const retrySleepTime = [2000, 5000, 10000];
+
+export async function retryAsyncFunction(func: ()=>Promise<any>, retryCount: number = 3) {
```
Comment (ua741) [inline src/utils/network/index.ts:5]: What would be the behavior when someone specify `retryCount  >=3`?
Also, if am not wrong, if we call the `retryAsyncFunction` with `retryCount=1`, we would end up retrying after 10000 unit of time..

File: src/utils/upload/index.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+export function segregateFiles(
+    filesWithCollectionToUpload: FileWithCollection[],
+) {
+    const metadataFiles: FileWithCollection[] = [];
+    const mediaFiles: FileWithCollection[] = [];
+    filesWithCollectionToUpload.forEach((fileWithCollection) => {
+        const file = fileWithCollection.file;
+        if (file?.name.substr(0, 1) === '.') {
```
Comment (ua741) [inline src/utils/upload/index.ts:40]: nit: can use `startsWith('.')` for better readability.

File: src/services/upload/uploader.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        if (e.message === CustomError.ETAG_MISSING) {
+            UIService.setFileProgress(rawFile.name, FileUploadResults.BLOCKED);
+            return { fileUploadResult: FileUploadResults.BLOCKED };
+        } else {
+            UIService.setFileProgress(rawFile.name, FileUploadResults.FAILED);
+            return { fileUploadResult: FileUploadResults.FAILED };
+        }
+    } finally {
+        file = null;
+        encryptedFile = null;
```
Comment (ua741) [inline src/services/upload/uploader.ts:81]: Both of these variables are defined within the scope of this function. Do we always need to set them as null for garbage collection? I am not very familiar with memory management in JS, so curious about this.

File: src/services/upload/uploader.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    UIService.setFileProgress(rawFile.name, 0);
+    let file: FileInMemory = null;
+    let encryptedFile: EncryptedFile = null;
+    try {
+        file = await UploadService.readFile(reader, rawFile, collection);
+
+        if (fileAlreadyInCollection(existingFilesInCollection, file)) {
+            // set progress to -2 indicating that file upload was skipped
+            UIService.setFileProgress(rawFile.name, FileUploadResults.SKIPPED);
+            await sleep(TwoSecondInMillSeconds);
```
Comment (ua741) [inline src/services/upload/uploader.ts:40]: What's the reason for adding the 2 second sleep here? Can we add a comment mentioning the same?

File: src/pages/gallery/index.tsx [hunk 6]
```diff
@@ -225,6 +225,7 @@ export default function Gallery() {
                         title: constants.SESSION_EXPIRED,
                         content: constants.SESSION_EXPIRED_MESSAGE,
                         staticBackdrop: true,
+                        nonClosable: true,
```
Comment (ua741) [inline src/pages/gallery/index.tsx:228]: Assuming you meant to make this change. Doesn't look related to Upload refactoring and linting..

File: src/services/upload/thumbnailService.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
+import { FILE_TYPE } from 'pages/gallery';
```
Comment (ua741) [inline src/services/upload/thumbnailService.ts:1]: Shall we also migrate `FILE_TYPE` enum out of `pages` namespace?

File: src/pages/gallery/index.tsx (hunk 6 above)
Comment (Abhinav-grd) [reply src/pages/gallery/index.tsx:228]: yes..i should move this to master, found it while making refactoring...and commit on this branch by mistake

File: src/utils/upload/index.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+            // ignore files with name starting with . (hidden files)
+            return;
+        }
+        if (file.name.slice(-4) === TYPE_JSON) {
```
[the remark this answers is not in the document]
Comment (Abhinav-grd) [reply src/utils/upload/index.ts]: ? is not required.  will remove it

File: src/utils/network/index.ts (hunk 2 above)
Comment (Abhinav-grd) [reply src/utils/network/index.ts]: so we never pass the retryCount and it always takes value as three, it is set so that next recursive call could be made with retryCount-1 value.

Comment (Abhinav-grd) [reply src/utils/network/index.ts]: will add a wrapper function around it to remove the confusion

File: src/services/upload/uploader.ts (hunk 4 above)
Comment (Abhinav-grd) [reply src/services/upload/uploader.ts:81]: there was a memory leak issue with uploader once and this solved it :sweat_smile: i think because of the recursive call made

File: src/services/upload/uploader.ts (hunk 5 above)
Comment (Abhinav-grd) [reply src/services/upload/uploader.ts:40]: so that file remain in the progressbar screen for 2 second with the message - "duplicate file skipping..."

File: src/services/upload/readFileService.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    ) {
+        fileType = FILE_TYPE.IMAGE;
+    }
+    return fileType;
+}
+
+export function getFileOriginalName(file: globalThis.File) {
+    let originalName: string = null;
+
+    const isEditedFile = file.name.endsWith(EDITED_FILE_SUFFIX);
```
Comment (ua741) [inline src/services/upload/readFileService.ts:42]: Is this some common standard where the edited file.name will be something like _imageName.jpeg-edited_ ? 
Can we add some comments/documentation about this part :)

File: src/services/upload/readFileService.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    if (isEditedFile) {
+        originalName = file.name.slice(0, -1 * EDITED_FILE_SUFFIX.length);
+    } else {
+        originalName = file.name;
+    }
+    return originalName;
+}
+
+function getFileStream(reader: FileReader, file: globalThis.File) {
```
Comment (ua741) [inline src/services/upload/readFileService.ts:51]: nit: We can it more generic and pass the `chunkSize` as a input to `getFileStream` and `fileChunkReaderMaker` method. This two func doesn't need to be aware of `ENCRYPTION_CHUNK_SIZE`.

File: src/services/upload/uploadService.ts [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+                extractedMetadata[key] = value;
+            }
+
+            const filedata = await getFileData(reader, rawFile);
+
+            return {
+                filedata,
+                thumbnail,
+                metadata: extractedMetadata,
+            } as FileInMemory;
```
Comment (ua741) [inline src/services/upload/uploadService.ts:135]: nit: should we also declare the `returnType` in `readFile` method signature?

File: src/services/upload/uploader.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        );
+
+        const uploadFile: UploadFile = UploadService.getUploadFile(
+            collection,
+            backupedFile,
+            encryptedFile.fileKey
+        );
+
+        const uploadedFile = await NetworkClient.uploadFile(uploadFile);
+        const decryptedFile = await decryptFile(uploadedFile, collection);
```
Comment (ua741) [inline src/services/upload/uploader.ts:62]: What's the purpose of decrypting the file right after encrypting and uploading. Is it possible to reuse the `file` which is passed in the `UploadService.encryptFile`.  I am curious about the impact of this method call during upload of large video lines.

File: src/services/upload/thumbnailService.ts [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+        } catch (e) {
+            logError(e);
+            // ignore and set staticThumbnail
+            hasStaticThumbnail = true;
+        }
+        const thumbnailBlob=await thumbnailCanvasToBlob(canvas);
+        const thumbnail = await getUint8ArrayView(
+            reader,
+            thumbnailBlob,
+        );
```
[the remark this answers is not in the document]
Comment (Abhinav-grd) [reply src/services/upload/thumbnailService.ts]: okk yaa..i should intialize the canvas with an empty canvas element, 
as generating a fallback thumbnail from blank canvas which is just a complete black screen

File: src/services/upload/thumbnailService.ts [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+export async function thumbnailCanvasToBlob(canvas:HTMLCanvasElement) {
+    let thumbnailBlob = null;
+    let attempts = 0;
+    let quality = 1;
+
+    do {
+        attempts++;
+        quality /= 2;
```
[the remark this answers is not in the document]
Comment (Abhinav-grd) [reply src/services/upload/thumbnailService.ts:159]: yes because the generated after the first attempt its size may not less then the desired thumbnailSize and then we attempt to reduce its quality more

File: src/services/upload/s3Service.ts [hunk 15]
```diff
@@ -0,0 +1,103 @@
+import { CHUNKS_COMBINED_FOR_A_UPLOAD_PART, DataStream } from './uploadService';
```
[the remark this answers is not in the document]
Comment (Abhinav-grd) [reply src/services/upload/s3Service.ts:1]: yes. i should rename it

File: src/services/upload/s3Service.ts [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+            uploadChunk,
+            progressTracker,
+        );
+        partEtags.push({ PartNumber: index + 1, ETag: eTag });
+    }
+    await completeMultipartUpload(partEtags, multipartUploadURLs.completeURL);
+    return multipartUploadURLs.objectKey;
+}
+
+export function getRandomProgressPerPartUpload(uploadPartCount: number) {
```
[the remark this answers is not in the document]
Comment (Abhinav-grd) [reply src/services/upload/s3Service.ts]: No..had forgotten to remove the export after refactoring...will remove

File: src/services/upload/readFileService.ts [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+
+export async function getFileData(reader: FileReader, file: globalThis.File) {
+    return file.size > MIN_STREAM_FILE_SIZE
+        ? getFileStream(reader, file)
+        : await getUint8ArrayView(reader, file);
+}
+
+export function getFileType(receivedFile: globalThis.File) {
+    let fileType: FILE_TYPE;
+    switch (receivedFile.type.split('/')[0]) {
```
[the remark this answers is not in the document]
Comment (Abhinav-grd) [reply src/services/upload/readFileService.ts:19]: ohh will add a helper function

File: src/services/upload/readFileService.ts (hunk 9 above)
Comment (Abhinav-grd) [reply src/services/upload/readFileService.ts:46]: okk..Will add comment :smile:

File: src/services/upload/uploader.ts (hunk 12 above)
Comment (Abhinav-grd) [reply src/services/upload/uploader.ts:62]: So we store the decrypted file Key along with file in localDB  ,and we never receive the actual file Key back in the uploader function so, we have to decrypt it using the collection key. we could have avoided decrypting the metadata but already had a util function for decryption so used it. Decrpyting metadata is cheap operation

So we are not decrypting the actual file just the metadata of the file so it should not be a problem.

Comment (ua741) [reply src/services/upload/uploader.ts:62]: Got it. I got confused. I was not aware that `uploadedFile` actually refers to `enteFile` object, not the media file.

File: src/services/upload/uploadService.ts [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+    getFileData,
+} from './readFileService';
+import { encryptFiledata } from './encryptionService';
+import { ENCRYPTION_CHUNK_SIZE } from 'types';
+import { uploadStreamUsingMultipart } from './multiPartUploadService';
+import UIService from './uiService';
+import { parseError } from 'utils/common/errorUtil';
+import { MetadataMap } from './uploadManager';
+
+export const MIN_STREAM_FILE_SIZE = 20 * 1024 * 1024;
```
Comment (ua741) [inline src/services/upload/uploadService.ts:23]: MULTIPART_PART_SIZE  
This is the chunk size of a file which will be uploaded in a single part.

Comment (ua741) [reply src/services/upload/uploadService.ts:23]: FILE_READER_CHUNK_SIZE = ENCRYPTION_CHUNK_SIZE

File: src/services/upload/multiPartUploadService.ts [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+    return percentPerPart;
+}
+
+async function combineChunksToFormUploadPart(
+    streamReader: ReadableStreamDefaultReader<Uint8Array>
+) {
+    const combinedChunks = [];
+    for (let i = 0; i < CHUNKS_COMBINED_FOR_A_UPLOAD_PART; i++) {
+        const { done, value: chunk } = await streamReader.read();
+        if (done) {
```
Comment (ua741) [inline src/services/upload/multiPartUploadService.ts:83]: Just to be safe, similar to streamReader.read(), shall we also return `{Uint8Array.from(combinedChunks), done}`

We should only call  `await completeMultipartUpload(partEtags, multipartUploadURLs.completeURL);` when we are 100% sure that we have read everything from `streamReader`

**Review by ua741 — APPROVED** [review]

File: src/services/upload/uploader.ts (hunk 12 above)
Comment (Abhinav-grd) [reply src/services/upload/uploader.ts:62]: will make this change on different branch will change confusion here

File: src/services/upload/exifService.ts [hunk 20]
```diff
@@ -0,0 +1,100 @@
+import EXIF from 'exif-js';
+
+import { logError } from 'utils/sentry';
+import { NULL_LOCATION, Location } from './metadataService';
+
+const SOUTH_DIRECTION = 'S';
+const WEST_DIRECTION = 'W';
+const EXIF_HAVING_TYPES = new Set(['jpeg', 'jpg', 'tiff']);
```
Comment (ua741) [inline src/services/upload/exifService.ts:8]: How do we handle exif data for HEIC file types? 
Also, is this list based on our observation or it's standard list? (Want to check that we are not missing out any other common types here)
